lib/logstash/outputs/dynatrace.rb

```diff
@@ -73,12 +71,44 @@ def headers
       def multi_receive(events)
         return if events.length.zero?
 
-        request = Net::HTTP::Post.new(uri, headers)
-        request.body = "#{LogStash::Json.dump(events.map(&:to_hash)).chomp}\n"
-        response = @client.request(request)
-        return if response.is_a? Net::HTTPSuccess
+        retries = 0
+        begin
+          request = Net::HTTP::Post.new(uri, headers)
+          request.body = "#{LogStash::Json.dump(events.map(&:to_hash)).chomp}\n"
+          response = send(request)
+          return if response.is_a? Net::HTTPSuccess
+
+          failure_message = "Dynatrace returned #{response.code} #{response.message}."
+
+          if response.is_a? Net::HTTPServerError
+            raise RetryableError.new failure_message
+          end
+
+          if response.is_a? Net::HTTPBadRequest
+            @logger.error("#{failure_message} Incorrectly formatted message.")
```

Comment:
```suggestion
            @logger.error(failure_message)
```
I'd just print the returned error message here which should include a reason and might also be about something other than the formatting. With https://github.com/dynatrace-oss/logstash-output-dynatrace/pull/8#discussion_r666175754 I just wanted to avoid having the "check if enabled and token permissions" guidance for 400s since that would've been misleading.
Alternatively use the guidance below only for 404s and for all other 4xx only the returned message.